presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedHistogramState.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator.aggregation.histogram;
+
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import org.openjdk.jol.info.ClassLayout;
+
+import static it.unimi.dsi.fastutil.HashCommon.arraySize;
+
+/**
+ * state object that uses a single histogram for all groups. See {@link GroupedTypedHistogram}
+ */
+public class GroupedHistogramState
+        extends AbstractGroupedAccumulatorState
+        implements HistogramState
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedHistogramState.class).instanceSize();
+    private TypedHistogram typedHistogram = null;
+    private long size;
+    private BlockBuilder valuesBlockBuilder;
+
+    public GroupedHistogramState(
+            Type keyType,
+            int expectedEntriesCount)
+    {
+        valuesBlockBuilder = keyType.createBlockBuilder(null, computeBucketCount(expectedEntriesCount, GroupedTypedHistogram.MAX_FILL_RATIO));
+        typedHistogram = new GroupedTypedHistogram(keyType, expectedEntriesCount, valuesBlockBuilder);
+    }
+
+    @Override
+    public void ensureCapacity(long size)
+    {
+        typedHistogram.ensureCapacity(size);
+    }
+
+    @Override
+    public TypedHistogram get()
+    {
+        return typedHistogram.setGroupId(getGroupId());
+    }
+
+    @Override
+    public void set(TypedHistogram value)
+    {
+        if (value instanceof GroupedTypedHistogram) {
+            if (typedHistogram != null) {
```

Comment:
i left a "checkState" (requireNonNull)  that the base histogram was non-null since we require non-null initialization; we also check the argument.

it's s dev error to get into a state where either is null and NOT a GroupedTypedHistogram.class because we don't support cross-merging implementations. in one query, it will be one type (grouped or single)

Reply:
done